backends/tc/runtime/pna.h

```diff
@@ -0,0 +1,235 @@
+static const ParserError_t ParserTimeout = 5;    // Parser execution time limit exceeded.
+static const ParserError_t ParserInvalidArgument = 6; // Parser operation was called with a value
+                                                      // not supported by the implementation
+
+enum PNA_Source_t { FROM_HOST, FROM_NET };
+
+enum PNA_MeterColor_t { RED, GREEN, YELLOW };
+
+enum MirrorType { NO_MIRROR, PRE_MODIFY, POST_MODIFY };
+
+struct pna_main_parser_input_metadata_t {
+    bool                     recirculated;
+    PortId_t                 input_port;    // taken from xdp_md or __sk_buff
+} __attribute__((aligned(4)));
+
+struct pna_main_input_metadata_t {
+    // All of these values are initialized by the architecture before
+    // the control block begins executing.
+    bool                     recirculated;
+    Timestamp_t              timestamp;         // taken from bpf helper
+    ParserError_t            parser_error;      // local to parser
+    ClassOfService_t         class_of_service;  // 0, set in control as global metadata
+    PortId_t                 input_port;
+} __attribute__((aligned(4)));;
+
+struct pna_main_output_metadata_t {
+    // The comment after each field specifies its initial value when the
+    // control block begins executing.
+    ClassOfService_t         class_of_service;
+} __attribute__((aligned(4)));
+
+/*
+ * Opaque struct to be used to share global PNA metadata fields.
+ * The size of this struct must be less than 32 bytes.
+ */
+struct pna_global_metadata {
+    bool             recirculated;
+    bool             drop; // NOTE : no drop field in PNA metadata, so we keep drop state as internal metadata.
+    PortId_t         egress_port;
+    enum MirrorType  mirror_type;
+    MirrorSlotId_t   mirror_slot_id;
+    MirrorSessionId_t mirror_session_id;
+    __u8             mark;
+    bool             pass_to_kernel; // internal metadata, forces sending packet up to kernel stack
+} __attribute__((aligned(4)));
+
+struct clone_session_entry {
+    __u32 egress_port;
+    __u16 instance;
+    __u8  class_of_service;
+    __u8  truncate;
+    __u16 packet_length_bytes;
+} __attribute__((aligned(4)));
+
+#define send_to_port(x) (compiler_meta__->egress_port = x)
+#define drop_packet() (compiler_meta__->drop = true)
+
+// structures and functions for tc backend
+
+struct p4tc_table_entry_act_bpf_params__local {
+    u32 pipeid;
+    u32 tblid;
+} __attribute__((preserve_access_index));
+
+struct p4tc_table_entry_act_bpf {
+    u32 act_id;
+    u8 params[124];
+};
+
+extern struct p4tc_table_entry_act_bpf *bpf_skb_p4tc_tbl_lookup(
+    struct __sk_buff *skb, struct p4tc_table_entry_act_bpf_params__local *params, void *key,
+    const u32 key__sz) __ksym;
+extern struct p4tc_table_entry_act_bpf *bpf_xdp_p4tc_tbl_lookup(
+    struct xdp_md *skb, struct p4tc_table_entry_act_bpf_params__local *params, void *key,
+    const u32 key__sz) __ksym;
+
+// port direction externs
```

Comment:
This looks irrelevant here and the next line too

Reply:
removed irrelevant comments